### PR TITLE
TCPS support in Single Instance Database

### PR DESCRIPTION
**Pull request by abhisbyk**
TCPS support (version 19.3.0 onwards). Self-signed certs generated from `orapki` utility is used to setup the wallets.
Only server auth is supported (no mTLS).

### CODE REVIEW

**Review by yunus-qureshi — APPROVED** [review]
Thanks

File: OracleDatabase/SingleInstance/README.md [hunk 1]
```diff
@@ -189,25 +189,22 @@ There are two ways to enable TCPS in the database:
...
+Similarly, to disable TCPS in the database, please use the following command:
 ```bash
     # Disable TCPS in the database
     docker exec -it <container name> /opt/oracle/configTcps.sh disable
 ```
 
 **NOTE**:
-- The container port at which TCPS listener is listening should be exposed to the outside world for access.
+- The container port at which TCPS listener is listening (i.e. 1522) should be exposed and mapped to some host port using `-p <host-port>:1522` option in the `docker run` command. It is required to connect to the database from the outside world using TCPS.
 - When TCPS is enabled, a self-signed certificate will be created for the same. For user's convenience, a client side wallet is prepared and stored at the location, `/opt/oracle/oradata/clientWallet/$ORACLE_SID`. You can use this client wallet along with `SQLPlus` to connect with the database. The sample command to download the client wallet is as follows:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:208]: Use the  correct product name, which is 'SQL\*Plus'  :)

File: OracleDatabase/SingleInstance/README.md [hunk 2]
```diff
@@ -226,7 +223,7 @@ Similarly, to disable the TCPS in the database, please use the following command
         docker exec -it <container name> /opt/oracle/configTcps.sh
     ```
     After certificate renewal, the client wallet should be updated by downloading it again.
-- Supports 21.3.0 XE also.
+- Supports XE version 21.3.0 onwards.
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:226]: Can you use 'Oracle Database XE' here?

**Review by cjbj — CHANGES_REQUESTED** [review]

**Comment by cjbj** [comment]
It could be helpful / clearer for users if you included the factoid from the PR description 'Only server auth is supported (no mTLS).' in the README text?

File: OracleDatabase/SingleInstance/README.md (hunk 1 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 2 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

**Comment by abhisbyk** [comment]
> It could be helpful / clearer for users if you included the factoid from the PR description 'Only server auth is supported (no mTLS).' in the README text?

Added this description in the Readme also. Thanks.

**Review by yunus-qureshi — DISMISSED** [review]

File: OracleDatabase/SingleInstance/README.md [hunk 3]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+Similarly, to disable TCPS in the database, please use the following command:
+```bash
+    # Disable TCPS in the database
+    docker exec -it <container name> /opt/oracle/configTcps.sh disable
+```
+
+**NOTE**:
+- Only the database server authentication is supported (no mTLS).
+- The container port at which TCPS listener is listening (i.e. 1522) should be exposed and mapped to some host port using `-p <host-port>:1522` option in the `docker run` command. It is required to connect to the database from the outside world using TCPS.
+- When TCPS is enabled, a self-signed certificate will be created for the same. For user's convenience, a client side wallet is prepared and stored at the location, `/opt/oracle/oradata/clientWallet/$ORACLE_SID`. You can use this client wallet along with `SQL*Plus` to connect with the database. The sample command to download the client wallet is as follows:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:209]: Wrong font?  Try without the code format but put a backslash before the asterisk: SQL\*Plus

File: OracleDatabase/SingleInstance/README.md [hunk 4]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+    ```bash
+        # ORACLE_SID default value is ORCLCDB
+        docker cp <container name>:/opt/oracle/oradata/clientWallet/<ORACLE_SID> <destination directory>
+    ```
+- The client wallet above will have wallet files, along with sample `sqlnet.ora` and `tnsnames.ora`. You should edit the `HOST` and `PORT` field accordingly in the `tnsnames.ora` before connecting using TCPS.
+- After the `tnsnames.ora` is modified, go inside the downloaded client wallet directory and set the TNS_ADMIN for `SQL*Plus` by using the `export TNS_ADMIN=$(pwd)` command. Then user can connect with the following sample commands:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:215]: Wrong font?  Try without the code format but put a backslash before the asterisk: SQL\*Plus

File: OracleDatabase/SingleInstance/README.md [hunk 5]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+    ```bash
+    # Connecting Enterprise Edition
+    sqlplus sys/<password>@ORCLCDB as sysdba
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:218]: Never encourage users to put passwords on the command line, particularly for SYS.  SQL*Plus will prompt for the password if it isn't given.

File: OracleDatabase/SingleInstance/README.md [hunk 6]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+    # Connecting Express Edition
+    sqlplus sys/<password>@XE as sysdba
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:220]: Never encourage users to put passwords on the command line, particularly for SYS.  SQL*Plus will prompt for the password if it isn't given.

**Review by cjbj — CHANGES_REQUESTED** [review]

File: OracleDatabase/SingleInstance/README.md [hunk 7]
```diff
@@ -341,6 +341,8 @@ docker run --name <container-name> -e CLONE_DB=true \
 -e PRIMARY_DB_CONN_STR='<the on-prem database connection string in <HOST>:<PORT>/<SERVICE_NAME> format>' \
 oracle/database:19-onprem
 ```
+**NOTE:**
+Make sure that the directory structure of the on-premise database matches with the directory structure used in the docker image.
```
Comment (Djelibeybi) [inline OracleDatabase/SingleInstance/README.md:345]:
```suggestion
Make sure that the directory structure of the on-premise database matches with the directory structure used in the container image.
```

File: OracleDatabase/SingleInstance/README.md (hunk 3 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 4 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 6 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 5 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 7 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

**Review by Djelibeybi — DISMISSED** [review]

File: OracleDatabase/SingleInstance/README.md [hunk 8]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
 In case this parameter is set `true` and passed to `docker run` command while reusing existing datafiles, even though this parameter would be visible as set to `true` in the container environment, this would not be set inside the database. The value used at the time of database creation will be used.
 
+#### Configuring TCPS in the database (Support from version 19.3.0 onwards)
+There are two ways to enable TCPS in the database:
+1. Enable TCPS while creating the database.
+2. Enable TCPS after the database is created.
+
+To enable TCPS while creating the database, use `-e ENABLE_TCPS=true` option with the `docker run` command. A listener endpoint will be created at the container port 1522 for TCPS.
+
+To enable TCPS after the database is created, please use one the following sample command:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:194]: Missing 'of' -> 'one of the'

File: OracleDatabase/SingleInstance/README.md [hunk 9]
```diff
@@ -78,7 +78,7 @@ After setting these environment variables, the container image can be built usin
 To run your Oracle Database image use the `docker run` command as follows:
 
     docker run --name <container name> \
-    -p <host port>:1521 -p <host port>:5500 \
+    -p <host port>:1521 -p <host port>:5500 -p <host port>: 1522\
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:81]: Is the space after the colon deliberate?

File: OracleDatabase/SingleInstance/README.md [hunk 10]
```diff
@@ -87,13 +87,14 @@ To run your Oracle Database image use the `docker run` command as follows:
...
     -e ENABLE_ARCHIVELOG=true \
+    -e ENABLE_TCPS=true \
     -v [<host mount point>:]/opt/oracle/oradata \
     oracle/database:21.3.0-ee
     
     Parameters:
        --name:        The name of the container (default: auto generated).
        -p:            The port mapping of the host port to the container port.
-                      Two ports are exposed: 1521 (Oracle Listener), 5500 (OEM Express).
+                      Two ports are exposed: 1521 (Oracle Listener), 5500 (OEM Express), 1522(TCPS Listener Port if TCPS is enabled).
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:97]: Add a space after 1522

File: OracleDatabase/SingleInstance/README.md [hunk 11]
```diff
@@ -118,6 +119,9 @@ To run your Oracle Database image use the `docker run` command as follows:
        -e ENABLE_ARCHIVELOG:
                       To enable archive log mode when creating the database (default: false).
                       Supported 19.3 onwards.
+       -e ENABLE_TCPS:
+                      To enable TCPS in the database.
+                      Supported 19.3 onwards.
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:124]: -> 'Supported by Oracle Database 19.3 onwards.'

File: OracleDatabase/SingleInstance/README.md [hunk 12]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
 
 In case this parameter is set `true` and passed to `docker run` command while reusing existing datafiles, even though this parameter would be visible as set to `true` in the container environment, this would not be set inside the database. The value used at the time of database creation will be used.
 
+#### Configuring TCPS in the database (Support from version 19.3.0 onwards)
+There are two ways to enable TCPS in the database:
+1. Enable TCPS while creating the database.
+2. Enable TCPS after the database is created.
+
+To enable TCPS while creating the database, use `-e ENABLE_TCPS=true` option with the `docker run` command. A listener endpoint will be created at the container port 1522 for TCPS.
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:192]: Change 'use' to 'use the'

File: OracleDatabase/SingleInstance/README.md [hunk 13]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+```
+
+Similarly, to disable TCPS in the database, please use the following command:
+```bash
+    # Disable TCPS in the database
+    docker exec -it <container name> /opt/oracle/configTcps.sh disable
+```
+
+**NOTE**:
+- Only the database server authentication is supported (no mTLS).
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:207]: Remove 'the'.

File: OracleDatabase/SingleInstance/README.md [hunk 14]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+- When TCPS is enabled, a self-signed certificate will be created for the same. For user's convenience, a client side wallet is prepared and stored at the location, `/opt/oracle/oradata/clientWallet/$ORACLE_SID`. You can use this client wallet along with SQL\*Plus to connect with the database. The sample command to download the client wallet is as follows:
+    ```bash
+        # ORACLE_SID default value is ORCLCDB
+        docker cp <container name>:/opt/oracle/oradata/clientWallet/<ORACLE_SID> <destination directory>
+    ```
+- The client wallet above will have wallet files, along with sample `sqlnet.ora` and `tnsnames.ora`. You should edit the `HOST` and `PORT` field accordingly in the `tnsnames.ora` before connecting using TCPS.
+- After the `tnsnames.ora` is modified, go inside the downloaded client wallet directory and set the TNS_ADMIN for SQL\*Plus by using the `export TNS_ADMIN=$(pwd)` command. Then user can connect with the following sample commands:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:215]: Change 'After the' to 'After' and 'the TNS_ADMIN' to 'TNS_ADMIN'.

File: OracleDatabase/SingleInstance/README.md [hunk 15]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+    ```bash
+    # Connecting Enterprise Edition
+    sqlplus sys@ORCLCDB as sysdba
+    # Connecting Express Edition
+    sqlplus sys@XE as sysdba
+    ```
+- The certificate used with TCPS has validity for 3 years. After certificate is expired, you can renew certs using the following command:
```
Comment (cjbj) [inline OracleDatabase/SingleInstance/README.md:222]: Chang 'After' to 'After the'.  

And probably change 'certs' to 'certifcates'?

**Review by cjbj — DISMISSED** [review]
Here are some wording fixes.  I don't need to re-review.

File: OracleDatabase/SingleInstance/README.md (hunk 8 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md:194]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 9 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: It was not deliberate. Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 10 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 11 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 12 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 13 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 14 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 15 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

**Comment by abhisbyk** [comment]
@gvenzl Please review and merge this PR. Thanks

**Review by yunus-qureshi — APPROVED** [review]

File: OracleDatabase/SingleInstance/README.md [hunk 16]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
 
 In case this parameter is set `true` and passed to `docker run` command while reusing existing datafiles, even though this parameter would be visible as set to `true` in the container environment, this would not be set inside the database. The value used at the time of database creation will be used.
 
+#### Configuring TCPS in the database (Support from version 19.3.0 onwards)
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:187]: Configuring TCPS connections for Oracle Database

File: OracleDatabase/SingleInstance/README.md [hunk 17]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+There are two ways to enable TCPS in the database:
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:188]: There are two ways to enable TCPS connections for the database

File: OracleDatabase/SingleInstance/README.md [hunk 18]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+1. Enable TCPS while creating the database.
+2. Enable TCPS after the database is created.
+
+To enable TCPS while creating the database, use the `-e ENABLE_TCPS=true` option with the `docker run` command. A listener endpoint will be created at the container port 1522 for TCPS.
+
+To enable TCPS after the database is created, please use one of the following sample commands:
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:194]: To enable TCPS after the database is created, please use the following sample command:

File: OracleDatabase/SingleInstance/README.md [hunk 19]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+```bash
+    # Creates Listener for TCPS at container port 1522
+    docker exec -it <container name> /opt/oracle/configTcps.sh
+```
+
+Similarly, to disable TCPS in the database, please use the following command:
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:200]: Similarly, to disable TCPS connections for the database

File: OracleDatabase/SingleInstance/README.md [hunk 20]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+```bash
+    # Disable TCPS in the database
+    docker exec -it <container name> /opt/oracle/configTcps.sh disable
+```
+
+**NOTE**:
+- Only database server authentication is supported (no mTLS).
+- The container port at which TCPS listener is listening (i.e. 1522) should be exposed and mapped to some host port using `-p <host-port>:1522` option in the `docker run` command. It is required to connect to the database from the outside world using TCPS.
+- When TCPS is enabled, a self-signed certificate will be created for the same. For user's convenience, a client side wallet is prepared and stored at the location, `/opt/oracle/oradata/clientWallet/$ORACLE_SID`. You can use this client wallet along with SQL\*Plus to connect with the database. The sample command to download the client wallet is as follows:
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:209]: - When TCPS is enabled, a self-signed certificate will be created. For users' convenience, a client-side wallet is prepared and stored at the location `/opt/oracle/oradata/clientWallet/$ORACLE_SID`. You can use this client wallet along with SQL\*Plus to connect to the database. The sample command to download the client wallet is as follows:

File: OracleDatabase/SingleInstance/README.md [hunk 21]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+    ```bash
+        # ORACLE_SID default value is ORCLCDB
+        docker cp <container name>:/opt/oracle/oradata/clientWallet/<ORACLE_SID> <destination directory>
+    ```
+- The client wallet above will have wallet files, along with sample `sqlnet.ora` and `tnsnames.ora`. You should edit the `HOST` and `PORT` field accordingly in the `tnsnames.ora` before connecting using TCPS.
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:214]: The client wallet directory above will include wallet files, along with sample `sqlnet.ora` and `tnsnames.ora` files. You should edit the `HOST` and `PORT` fields accordingly in the `tnsnames.ora` before connecting using TCPS.

File: OracleDatabase/SingleInstance/README.md [hunk 22]
```diff
@@ -180,6 +184,49 @@ Archive mode can be enabled during the first time when database is created by se
...
+- After `tnsnames.ora` is modified, go inside the downloaded client wallet directory and set TNS_ADMIN for SQL\*Plus by using the `export TNS_ADMIN=$(pwd)` command. Then user can connect with the following sample commands:
```
Comment (gvenzl) [inline OracleDatabase/SingleInstance/README.md:215]: Then users can connect via TPCS with, for example, the following commands:

**Review by gvenzl — CHANGES_REQUESTED** [review]
There are a couple of minor comments/typos in the ReadMe to fix, please.

File: OracleDatabase/SingleInstance/README.md [hunk 23]
```diff
@@ -87,37 +87,41 @@ To run your Oracle Database image use the `docker run` command as follows:
...
     -e ENABLE_ARCHIVELOG=true \
+    -e ENABLE_TCPS=true \
     -v [<host mount point>:]/opt/oracle/oradata \
     oracle/database:21.3.0-ee
     
     Parameters:
        --name:        The name of the container (default: auto generated).
        -p:            The port mapping of the host port to the container port.
-                      Two ports are exposed: 1521 (Oracle Listener), 5500 (OEM Express).
+                      Two ports are exposed: 1521 (Oracle Listener), 5500 (OEM Express), 1522 (TCPS Listener Port if TCPS is enabled).
```
[the remark this answers is not in the document]
Comment (cjbj) [reply OracleDatabase/SingleInstance/README.md:97]: That should be "The following ports"

[the remark this answers is not in the document]
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md [hunk 24]
```diff
@@ -87,37 +87,41 @@ To run your Oracle Database image use the `docker run` command as follows:
...
-                      Supported 19.3 onwards.
+                      Supported by Oracle Database 19.3 onwards.
        -e ORACLE_CHARACTERSET:
                       The character set to use when creating the database (default: AL32UTF8).
        -e ENABLE_ARCHIVELOG:
                       To enable archive log mode when creating the database (default: false).
-                      Supported 19.3 onwards.
+                      Supported by Oracle Database 19.3 onwards.
+       -e ENABLE_TCPS:
+                      To enable TCPS in the database.
```
[the remark this answers is not in the document]
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 16 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 17 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 18 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 19 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 20 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 21 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

File: OracleDatabase/SingleInstance/README.md (hunk 22 above)
Comment (abhisbyk) [reply OracleDatabase/SingleInstance/README.md]: Fixed.

**Comment by abhisbyk** [comment]
> There are a couple of minor comments/typos in the ReadMe to fix, please.

Fixed the PR according to your suggestions. Thanks

**Review by gvenzl — APPROVED** [review]

**Pull request merged**